packages/react-core/src/components/Nav/NavItem.tsx

```diff
@@ -39,12 +54,148 @@ export const NavItem: React.FunctionComponent<NavItemProps> = ({
   preventDefault = false,
   onClick = null as NavSelectClickHandler,
   component = 'a',
+  flyout,
+  onShowFlyout,
   ouiaId,
   ouiaSafe,
   ...props
 }: NavItemProps) => {
-  const Component = component as any;
+  const { flyoutRef, setFlyoutRef, disableHover } = React.useContext(NavContext);
   const { isNavOpen } = React.useContext(PageSidebarContext);
+  const flyoutContext = React.useContext(FlyoutContext);
+  const [flyoutXDirection, setFlyoutXDirection] = React.useState(flyoutContext.direction);
+  const ref = React.useRef<HTMLLIElement>();
+  const flyoutVisible = ref === flyoutRef;
+  const Component = component as any;
+  const hasFlyout = flyout !== undefined;
+
+  const showFlyout = (show: boolean) => {
+    if (!flyoutVisible && show) {
+      setFlyoutRef(ref);
+    } else if (flyoutVisible && !show) {
+      setFlyoutRef(null);
+    }
+    onShowFlyout && show && onShowFlyout();
+  };
+
+  useIsomorphicLayoutEffect(() => {
+    if (hasFlyout && ref.current && canUseDOM) {
+      const flyoutMenu = ref.current.lastElementChild as HTMLElement;
+      if (flyoutMenu && flyoutMenu.classList.contains(styles.navSubnav)) {
+        const origin = ref.current.getClientRects()[0];
+        const rect = flyoutMenu.getClientRects()[0];
+        if (origin && rect) {
+          const spaceLeftLeft = origin.x - rect.width;
+          const spaceLeftRight = window.innerWidth - origin.x - origin.width - rect.width;
+          let xDir = flyoutXDirection as 'left' | 'right' | 'none';
+          if (spaceLeftRight < 0 && xDir !== 'left') {
+            setFlyoutXDirection('left');
+            xDir = 'left';
+          } else if (spaceLeftLeft < 0 && xDir !== 'right') {
+            setFlyoutXDirection('right');
+            xDir = 'right';
+          }
+          let xOffset = 0;
+          if (spaceLeftLeft < 0 && spaceLeftRight < 0) {
+            xOffset = xDir === 'right' ? -spaceLeftRight : -spaceLeftLeft;
+          }
+          if (xDir === 'left') {
+            // console.log("Something here isn't working");
+            flyoutMenu.classList.add(menuStyles.modifiers.left);
+            flyoutMenu.style.setProperty(rightOffset.name, `-${xOffset}px`);
+          } else {
+            flyoutMenu.style.setProperty(leftOffset.name, `-${xOffset}px`);
+          }
+
+          const spaceLeftBot = window.innerHeight - origin.y - rect.height;
+          const spaceLeftTop = window.innerHeight - rect.height;
+          if (spaceLeftTop < 0 && spaceLeftBot < 0) {
+            // working idea: page can usually scroll down, but not up
+            // TODO: proper scroll buttons
```

Comment:
do we need issue for this?

Reply:
This is also in our Menu flyouts, I don't know if we have a feature to track it yet